### PR TITLE
Enhance frictional models based on ExternalCombiTable1D

### LINKED ISSUE #3661: Improve table design of Modelica.Mechanics.Rotational.Components.Clutch to use static tables instead of dynamic
`Modelica.Mechanics.Rotational.Components.Clutch` uses `Modelica.Math.Vectors.interpolate()` for table interpolation. The used tables are static, i.e., never changed throughout simulation time. But vector-interpolation is meant for dynamic tables – i.e., tables whose content is constructed out of simulated runtime values. I consider this a design bug.

Even more, the parameter `mu0` is modeled as a variable. Its equation is
```
  // Constant auxiliary variable
  mu0 = Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], 0, 1);
```
Thus, the equation itself comments that it is not a real equation, but a constant. It will never change at runtime. It is a static interpolation using a static table.

The logic for deciding the clutch mode is now mingled with table interpolation to define the friction torque `tau`. The rationale for this design seems to be that the desired abscissa value for the interpolation (the sign for `w_rel`) depends on the clutch mode; and likewise the ordinate value interpolation result (the sign of `tau`) depends on the clutch mode:
```
tau = if locked then sa*unitTorque else if free then 0 else cgeo*fn*(
    if startForward then 
      Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], w_rel, 1)
    else if startBackward then 
      -Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], w_rel, 1)
    else if pre(mode) == Forward then 
      Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], w_rel, 1)
    else 
      -Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], -w_rel, 1));
```

I propose to change this design to ordinary `Modelica.Tables.Interpolation1Ds` combi-tables, leveraging on the fact that the used interpolation table is static. Doing so will help to generate faster simulation code and clarify the design, making it explicit that static table interpolation is used and due to that only certain values can change.

The following is an alternative design based on combi-tables:
```
model ClutchNoDynamicTable
  extends Modelica.Mechanics.Rotational.Icons.Clutch;
  extends Modelica.Mechanics.Rotational.Interfaces.PartialCompliantWithRelativeStates;
  extends Modelica.Mechanics.Rotational.Interfaces.PartialFriction;
  extends Modelica.Thermal.HeatTransfer.Interfaces.PartialElementaryConditionalHeatPortWithoutT;

  parameter Real mue_pos[:, 2] = [0, 0.5]
    "[w,mue] positive sliding friction coefficient (w_rel>=0)";
  parameter Real peak(final min = 1) = 1
    "Peak for maximum value of mue at w==0 (mue0_max = peak*mue_pos[1,2])";
  parameter Real cgeo(final min = 0) = 1
    "Geometry constant containing friction distribution assumption";
  parameter Modelica.SIunits.Force fn_max(final min = 0, start = 1)
    "Maximum normal force";

  Modelica.SIunits.Force fn
    "Normal force (fn=fn_max*f_normalized)";
  Modelica.Blocks.Interfaces.RealInput f_normalized
    "Normalized force signal 0..1 (normal force = fn_max*f_normalized; clutch is engaged if > 0)"
    annotation (Placement(transformation(
      origin = {0,110},
      extent = {{20,-20},{-20,20}},
      rotation = 90)));

protected 
  final parameter Real mue0 = Modelica.Math.Vectors.interpolate(
    mue_pos[:,1],
    mue_pos[:,2],
    0,
    1)
    "Friction coefficient for w=0 and forward sliding"
    annotation(Evaluate = true);

  final Real table_signs[2]
    "Signs for sliding friction coefficient table interpolation:
     1: sign for table.y
     2: sign for table.u = w_rel";
  final eFMI.Tables.CombiTable1Ds table(
    table = mue_pos)
    "Sliding friction coefficient table.";

equation 
  // Relative quantities
  w_relfric = w_rel;
  a_relfric = a_rel;

  // Normal force and friction torque for w_rel=0
  fn = fn_max*f_normalized;
  free = fn <= 0;
  tau0 = mue0*cgeo*fn;
  tau0_max = peak*tau0;

  // Friction torque
  table_signs =
    if startForward then 
      {1,1}
    elseif startBackward then 
      {-1,1}
    elseif pre(mode) == Forward then 
      {1,1}
    else 
      {-1,-1};
  table.u = table_signs[2] * w_rel;
  tau =
    if locked then sa*unitTorque
    elseif free then 0
    else cgeo*fn*(table_signs[1]*table.y[1]);
  lossPower = tau*w_relfric;

  end ClutchNoDynamicTable;
```

This design uses `annotation(Evaluate = true)` to statically evaluate `mue0`. It also separates the interpolation sign decisions based on the clutch mode (dynamic) from the interpolation in the static table itself:
```
// Friction torque
  table_signs =
    if startForward then 
      {1,1}
    elseif startBackward then 
      {-1,1}
    elseif pre(mode) == Forward then 
      {1,1}
    else 
      {-1,-1};
  table.u = table_signs[2] * w_rel;
  tau =
    if locked then sa*unitTorque
    elseif free then 0
    else cgeo*fn*(table_signs[1]*table.y[1]);
```
The `table_signs` vector encodes the different modes of harmonizing the interpolation point (`w_rel`) and interpolation result (`tau`) with the clutch mode.

I tested this redesign and it is functional equivalent, but non-functional more efficient (at least in Dymola). Considering that `Modelica.Math.Vectors.interpolate` is an in Modelica implemented function -- and not C code -- this should be not surprising. Note, that the clutch does not leverage on the last index used for interpolation (`iLast`) to gain any faster lookup (regardless if such would help at all in the first place considering combi-tables are using an efficient binary search to speed-up lookup).

### PR DESCRIPTION
**Pull request by beutlich**
@christoff-buerger This adds your proposal as PR. Note that I swapped the components of table_signs to match the columns of mu_pos.

Since CombiTable1Ds supports more interpolation smoothness kinds we also could get rid of

https://github.com/modelica/ModelicaStandardLibrary/blob/81dd1b63a61075c85fbd26945eb333b202b9e4b9/Modelica/Mechanics/Rotational/Components/Clutch.mo#L102-L103

, but might lose mu0 being a parameter again.

Closes #3661

### CODE REVIEW

File: Modelica/Mechanics/Rotational/Components/Clutch2.mo
Comment (tobolar) [inline Modelica/Mechanics/Rotational/Components/Clutch2.mo:51]: Possible to change to (for even better readability):
```
  table_signs =
    if     startForward         then { 1, 1}
    elseif startBackward        then { 1,-1}
    elseif pre(mode) == Forward then { 1, 1}
    else                             {-1,-1};
```

Comment (beutlich) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo:51]: Sure, go ahead.

File: Modelica/Mechanics/Rotational/Components/Clutch2.mo
Comment (beutlich) [inline Modelica/Mechanics/Rotational/Components/Clutch2.mo:51]:
```suggestion
  tau = if locked then sa*unitTorque elseif free then 0 else mu*cgeo*fn;
```

Have it similar as tau0.

**Comment by otronarp** [comment]
May a suggest a name change? Calling things `Foo2` is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

**Comment by beutlich** [comment]
> 
> 
> May I suggest a name change? Calling things `Foo2` is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

I only did so for the sake of consistency since we alraedy have ElastoBacklash and ElastoBacklash2 in Rotational.Components (and even other component models like Diode and Diode2).

**Comment by tobolar** [comment]
> It is an slightly improved implementation.

Actually, this reimplementation shall have the same functionality as the original clutch model. So why is it necessary to double the it?

**Comment by tobolar** [comment]
We shall also reimplement `Mechanics.Rotational.Components.Brake` accordingly.

**Review by MartinOtter — CHANGES_REQUESTED** [review]
Very good proposal (the "current" friction elements had been implemented at a time, where no table interpolation was available and therefore a quick implementation with a Modelica function was made;).

However, it would be good to use the new friction elements at least in one example (maybe I missed this).

**Comment by beutlich** [comment]
> 
> However, it would be good to use the new friction elements at least in one example (maybe I missed this).

I already updated Modelica.Mechanics.Rotational.Examples.CoupledClutches to utilize Clutch2.

> Currently, only linear interpolation in the table is supported.

@MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.

**Comment by MartinOtter** [comment]
> 
> 
> > However, it would be good to use the new friction elements at least in one example (maybe I missed this).
> 
> I already updated Modelica.Mechanics.Rotational.Examples.CoupledClutches to utilize Clutch2.

O.k.

> 
> > Currently, only linear interpolation in the table is supported.
> 
> @MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.

Yes, smoothness makes case and mu0 should no longer be used with Vectors.interpolate.

**Comment by beutlich** [comment]
> > @MartinOtter Do you agree to also add the interpolation smoothness as new parameter? In that case we no longer should use Modelica.Math.Vectors.interpolate to determine mu0.
> 
> Yes, smoothness makes sense and mu0 should no longer be used with Vectors.interpolate.

Done for Clutch2 for first review. @tobolar If you agree we should migrate Brake2 and OneWayClutch2 from block CombiTable1D to external object ExternalCombiTable1D, too.

**Comment by beutlich** [comment]
> Actually, this reimplementation shall have the same functionality as the original clutch model. So why is it necessary to double the it?

I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason to not simply update the original models. @MartinOtter needs to decide here.

**Comment by christoff-buerger** [comment]
> Note that I swapped the components of table_signs to match the columns of mu_pos.

That is good (= more intuitive).

 > May I suggest a name change? Calling things Foo2 is not helpful to the user unless the two signifies that there are two of something within the model (and as far as I understand it that's not the case here, it's an alternative implementation right?).

 > It is an slightly improved implementation. I only did so for the sake of consistency since we already have ElastoBacklash and ElastoBacklash2 in Rotational.Components (and even other component models like Diode and Diode2).

I agree with @otronarp. We should avoid doubling and names like `Foo2`.

Is there any rationale or argument why we need two clutch implementations; why we have to keep the old as well? Does the redesign have any drawbacks (functional or non-functional like useabelility or performance)?

This is particularly important because

 > We shall also reimplement Mechanics.Rotational.Components.Brake and Mechanics.Rotational.Components.OneWayClutch accordingly.

which would result in a plethora of `Foo2` things.

Do we also double the test models/examples utilizing the redesign? Or would we just not use one of the two implementations?

Thus, I like to push on:

 > I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason (except that protected mu0 changed variability from variable to parameter) to not simply update the original models. @MartinOtter needs to decide here. If he agrees we can avoid the duplication of the models.

@MartinOtter what is your opinion?

**Comment by christoff-buerger** [comment]
@tobolar: The `Brake2` still uses:
```
 final parameter Real mu0 = Modelica.Math.Vectors.interpolate(mu_pos[:,1], mu_pos[:,2], 0, 1)
```
i.e., always linear segments as smoothness; likewise the `OneWayClutch2`.

I propose to handle it like @beutlich improved for `Clutch2`, i.e., to support different table smoothness.

**Comment by christoff-buerger** [comment]
Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

**Comment by beutlich** [comment]
> 
> 
> Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

Note, that I added Clutch2 and updated it with full table smoothnes just as an offer to discuss with the library officers to get their opinion which way to head. Once this is clarifief we can modify/add Brake and OneWayClutch accordingly.

**Comment by tobolar** [comment]
`Clutch2` looks fine to me now and the smoothness option is great.

Since the piece of code 
```
    if     smoothness == Types.Smoothness.ConstantSegments then TabInternal.getTable1DValueNoDer(tableID, 1, 0)
    elseif smoothness == Types.Smoothness.LinearSegments   then TabInternal.getTable1DValueNoDer2(tableID, 1, 0)
    else                                                        TabInternal.getTable1DValue(tableID, 1, 0)
```
will be used also in other elements (and they are not only Rotational, see #3664), we shall extract it into a separate function. What could be the best location for such a function?

File: Modelica/Mechanics/Rotational/Components/Clutch2.mo
Comment (beutlich) [inline Modelica/Mechanics/Rotational/Components/Clutch2.mo:7]: Renaming imports should be avoided. Better:

```mo
import Modelica.Blocks.Tables;
```

**Comment by beutlich** [comment]
> getTable1DValue

I also thought about this. My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) and calling the functions from Internal. @tobolar Can you do it?

**Comment by tobolar** [comment]
> Can you do it?

I will prepare a suggestion.

File: Modelica/Mechanics/Rotational/Components/Clutch2.mo
Comment (beutlich) [inline Modelica/Mechanics/Rotational/Components/Clutch2.mo:10]: I'd actually avoid to introduce cosmetical changes now. This should go to a separate issue addressing all component models.

Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo:10]: Ok, how to revert this?

Comment (beutlich) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo:10]: > 
> 
> Ok, how to revert this?

```
git reset ca8078fbd49f1f9ab892818bbcd93a3d60ac12f1
git push --force
```

**Review by MartinOtter — APPROVED** [review]

File: Modelica/Mechanics/Rotational/Components/Clutch2.mo
Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo]: Done. The first time I've ever did it!

Comment (beutlich) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo]: You still have the changes locally. So be careful when pushing again.

Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo]: I use TortoiseSVN which enables three options how to reset a commit. I selected the one where also local changes are reset.
But anyway, thanks for the hint because I've never done it so far.

Comment (dietmarw) [reply Modelica/Mechanics/Rotational/Components/Clutch2.mo]: I truly hope you are **not** using Tprotois**SVN** in order to work on the git repo ;-)

**Comment by tobolar** [comment]
@beutlich 
> My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) ...

Actually, the smoothness is already given in the constructor of ExternalCombiTable1D. Is it reasonable to have it as an additional argument then?

**Comment by beutlich** [comment]
> 
> 
> @beutlich
> 
> > My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) ...
> 
> Actually, the smoothness is already given in the constructor of ExternalCombiTable1D. Is it reasonable to have it as an additional argument then?

Yes, since we decide in the Modelica side which if the functions to call.

**Comment by HansOlsson** [comment]
> > What could be the best location for such a function?
> 
> I also thought about this. My proposal is to introduce inline function Modelica.Blocks.Tables.getTable1DValue (or Modelica.Blocks.Tables.Functions.getTable1DValue) (taking the smoothness as argument) and calling the functions from Internal. @tobolar Can you do it?

Note that inlining of this function is very important, as we want to be able to differentiate for some values of smoothness, but cannot for others. If we cannot guarantee that inline+optional differentiation doesn't work we shouldn't introduce the new function.

**Comment by tobolar** [comment]
Trying to implement the function. But having smoothness as an argument can lead to undesired results, see example below where parameters `smoothnessConstructor` and `smoothnessEvaluator` have different values. Or do I understand it wrong?

```
function getTable1DValue "Interpolate one-dimensional table defined by matrix"
  import Modelica.Blocks.Types.Smoothness;
  import Modelica.Blocks.Tables.Internal;

  input Modelica.Blocks.Types.ExternalCombiTable1D tableID;
  input Integer icol;
  input Smoothness smoothness = Smoothness.LinearSegments;
  input Real u;
  output Real y;
algorithm 
  y :=
    if     smoothness == Smoothness.ConstantSegments then Internal.getTable1DValueNoDer(tableID, icol, u)
    elseif smoothness == Smoothness.LinearSegments   then Internal.getTable1DValueNoDer2(tableID, icol, u)
    else                                                  Internal.getTable1DValue(tableID, icol, u);
end getTable1DValue;

model Test
  import Modelica.Blocks.Types;

  parameter Real table[:, 2] = [0, 0.5;1, 1; 2,3; 3,4; 4,2; 5,2.5; 6,3];
  parameter Types.Smoothness smoothnessConstructor = Types.Smoothness.LinearSegments;
  parameter Types.Smoothness smoothnessEvaluator = Types.Smoothness.ConstantSegments;
  Real y;
protected 
  final parameter Types.ExternalCombiTable1D tableID = Types.ExternalCombiTable1D(
    tableName="NoName",
    fileName="NoName",
    table=table,
    columns={2},
    smoothness=smoothnessConstructor,
    extrapolation=Types.Extrapolation.LastTwoPoints,
    verboseRead=false);
equation 
  y = getTable1DValue(tableID, 1, smoothnessEvaluator, time);
end Test;
```

**Comment by beutlich** [comment]
> Or do I understand it wrong?

I see your point even though it is not relevant for the current use case. Unfortunately we do not (yet) have a getter function of the external object to retrieve its ctor settings. To keep things simple I propose to leave the table value call as is, even if it means to write duplicate code and long lines. (It is in a protected section anyway.)

**Comment by beutlich** [comment]
> Above two things, (1) can we avoid doubling of clutch, brake etc models and (2) consistent support for full table smoothness, have to be addressed from my point of view before a positive review.

I refactored again and now there are no more duplicated classes.

**Comment by beutlich** [comment]
What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.

**Comment by beutlich** [comment]
Please also not note that chaning mu0 from public to protected section is a backward-compatibility changing break that needs at least be documented (in the release notes).

**Comment by tobolar** [comment]
> Please also not note that chaning mu0 from public to protected section is a backward-compatibility breaking change...

We can also make it public again and protect it in the next major version. But actually, I can hardly imagine mu0 is a matter of interesse of a user.

**Review by tobolar — APPROVED** [review]
Fine for me now.

**Review by tobolar — CHANGES_REQUESTED** [review]
BearingFriction and LossyGear seem to be affected as well. I can care.

**Comment by tobolar** [comment]
Now BearingFriction and LossyGear are changed as well.

In LossyGear, variable `ideal` can be changed to parameter as well.

**Comment by beutlich** [comment]
> In LossyGear, variable `ideal` can be changed to parameter as well.

`ideal` is an unnecessary optimization that can be removed. Also `interpolation_result` is not required at all, especially not as matrix.

**Comment by tobolar** [comment]
I agree. What about the backward compatibility.

File: Modelica/Mechanics/Rotational/Components/LossyGear.mo
Comment (HansOlsson) [inline Modelica/Mechanics/Rotational/Components/LossyGear.mo:29]: It is not clear to me that using a constant is better here. You now see `nCol-1` and also a description that describes **four** columns; so I'm more left wondering whether they are the same or not. As far as I understand there is no plan to make nCol variable. If we are going with such a constant it would to me be more logical to have nValueCols=4, and replace nCol and nCol-1 with nValueCols+1 and nValueCols; as it seems more normal to add columns than to remove them.

Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/LossyGear.mo:29]: The introduction of nCol is rather intended for consistency in dimension of lossTable and constructor of tableID. But it shall also be consistent for interpolation_result. So I simply took a short named constant for it.

I have no preference for it - we can also call it e.g. nValue=4.

Comment (beutlich) [reply Modelica/Mechanics/Rotational/Components/LossyGear.mo:29]: I propose to remove constant nCol again. It is protected, but SimulationX displays it as dimensions of lossTable where nCol cannot be deduced at all.

Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/LossyGear.mo:29]: Ok, I will revert the changes on it.

Comment (tobolar) [reply Modelica/Mechanics/Rotational/Components/LossyGear.mo]: Reverted with 8a2bb0a7a3d336c5b49537e3de01feef984d3b6d

**Review by tobolar — APPROVED** [review]
Fine for me.

**Comment by beutlich** [comment]
> 
> 
> What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.

@HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.

**Comment by HansOlsson** [comment]
> > What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.
> 
> @HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.

To me "Evaluate=true" should be removed as mu0 should not be evaluated during translation, since it depends on the external table (and evaluating that during translation requires quite specific solution and doesn't really help).

**Review by HansOlsson — CHANGES_REQUESTED** [review]
As indicated remove the Evaluate=true for mu0.

**Comment by beutlich** [comment]
> 
> 
> As indicated remove the Evaluate=true for mu0.

Done. Please update your review. Thank you again!

**Review by HansOlsson — APPROVED** [review]
Looks good now.

**Comment by beutlich** [comment]
OK, waiting for @christoff-buerger's final approval then.

**Comment by MartinOtter** [comment]
>    I agree that the revised models Clutch2/Brake2/OneWayClutch2 fall back to the same functionality as the legacy models Clutch/Brake/OneWayClutch. There is no particular reason (except that protected mu0 changed variability from variable to parameter) to not simply update the original models. @MartinOtter needs to decide here. If he agrees we can avoid the duplication of the models.

As long as the changes are backwards compatible (e.g. only new parameters added, but not old parameters removed or changed), it is of course better to have only one model and not two different models.

**Comment by christoff-buerger** [comment]
> 
> 
> > > What I still not understand is why mu0 gets the Evaluate=true annotation, but depends on mu_pos being default-evaluated. I didn't question it at first sight when @christoff-buerger suggested it, but would like to get rid of it again now that mu0 is of parameter variability anyway.
> > 
> > 
> > @HansOlsson @christoff-buerger Can you please confirm that the Evaluate annotation for mu0 should be removed again. Thanks.
> 
> To me "Evaluate=true" should be removed as mu0 should not be evaluated during translation, since it depends on the external table (and evaluating that during translation requires quite specific solution and doesn't really help).

The "Evaluate=true" stems from the original redesign where mu0 was still defined by means of `Modelica.Math.Vectors.interpolate` and not a combi-table.

Since we decided since than that full table interpolation modes should be supported (not just linear), we also have to define mu0 via a combi-table now (because we need to support spline interpolation & co).

From a design point, mu0 is a constant. It is pity that we cannot use "Evaluate=true" with it because combi-tables as external objects are challenging for Modelica tools. To that end, I see your concern as a tool shortcoming/challenge.

It is an important point however -- that tables are not suited for "Evaluate=true" -- such that this annotation should be dropped.

**Review by christoff-buerger — APPROVED** [review]
Looks good. We have all relevant classes redesigned to use tables instead of `Modelica.Math.Vectors.interpolate` and support all kind of table smoothness/interpolation modes.

There are no class / functionality duplications like `Clutch2` etc.

We also have no `Evaluate=true` on constants derived via table interpolation, since such is troublesome for external table objects (a _common_ tooling shortcoming).

**Comment by christoff-buerger** [comment]
It can be merged from my perspective.

**Pull request merged**
